In [1]:
# Importing pandas as pd
import pandas as pd

In [2]:
# Reading ExcelFile.
xls = pd.ExcelFile('Data.xlsx')

# Previewing sheet names inside excel file.
xls.sheet_names

['RH K-4 MOY Data',
 'MOY Export',
 'Syracuse',
 'Howard',
 'Swarthmore',
 'Spelman',
 'Williams',
 'Brooklyn',
 'Tuskegee',
 'Northwestern',
 'Georgetown',
 'Wisconsin']

In [3]:
# Combining all student records into a single Data Frame (df), excluding first two summary sheets
df = pd.concat(pd.read_excel(xls, sheet_name=xls.sheet_names[2:]), ignore_index=True)
print(df.shape)
df.head(15)

(274, 17)


Student_ Number   Grade_Level  Home_room  BOY Indep Level  \
0        236563003.0           0.0   Syracuse              AA   
1        238194344.0           0.0   Syracuse              AA   
2        239980246.0           0.0   Syracuse              AA   
3        240845834.0           0.0   Syracuse              AA   
4        239980196.0           0.0   Syracuse              AA   
5        240414854.0           0.0   Syracuse              AA   
6        240158048.0           0.0   Syracuse              AA   
7        240060129.0           0.0   Syracuse              AA   
8        240267807.0           0.0   Syracuse              AA   
9        236665122.0           0.0   Syracuse              AA   
10       239980253.0           0.0   Syracuse              AA   
11       240414904.0           0.0   Syracuse              AA   
12       238742233.0           0.0   Syracuse              AA   
13       239352479.0           0.0   Syracuse              AA   
14       239980238.0           0.0   Syracuse              AA   

     Grade Level Equiv  Inst Level  Grade Level Equiv.1 MOY Indep Level  \
0                -0.25         NaN                                   AA   
1                -0.25         NaN                                   AA   
2                -0.25         NaN                                   AA   
3                -0.25         NaN                                   AA   
4                -0.25         NaN                                   AA   
5                -0.25         NaN                                    A   
6                -0.25         NaN                                    A   
7                -0.25         NaN                                    A   
8                -0.25         NaN                                    A   
9                -0.25         NaN                                    A   
10               -0.25         NaN                                    A   
11               -0.25         NaN                                    B   
12               -0.25         NaN                                    B   
13               -0.25         NaN                                    B   
14               -0.25         NaN                                    B   

    Grade Level Equiv.2  Inst Level.1  Grade Level Equiv.3   Accuracy (%)  \
0                 -0.25             A                  0.0           88.0   
1                 -0.25             A                  0.0           97.0   
2                 -0.25             A                  0.0           33.0   
3                                  AA                -0.25            NaN   
4                                  AA                -0.25            NaN   
5                   0.0             B                 0.33           90.0   
6                   0.0             B                 0.33           95.0   
7                   0.0             B                 0.33           90.0   
8                   0.0             B                 0.33           90.0   
9                   0.0             B                 0.33           90.0   
10                                  B                 0.33           95.0   
11                 0.33             C                 0.67           90.0   
12                 0.33             C                 0.67           90.0   
13                 0.33             C                 0.67           90.0   
14                 0.33             C                 0.67           90.0   

     Fluency (0-3)   Comp Within (0-3)   Comp Beyond (0-3)  Comp About  \
0              1.0                 3.0                 2.0         NaN   
1              2.0                 2.0                 1.0         NaN   
2              2.0                 1.0                 2.0         NaN   
3              3.0                 1.0                 NaN         NaN   
4              2.0                 3.0                 1.0         NaN   
5              NaN                 1.0                 2.0         NaN   
6          

In [4]:
# previewing df
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 274 entries, 0 to 273
Data columns (total 17 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Student_ Number       274 non-null    float64
 1   Grade_Level           274 non-null    float64
 2   Home_room             274 non-null    object 
 3   BOY Indep Level       273 non-null    object 
 4    Grade Level Equiv    259 non-null    float64
 5    Inst Level           187 non-null    object 
 6    Grade Level Equiv.1  274 non-null    object 
 7   MOY Indep Level       271 non-null    object 
 8    Grade Level Equiv.2  274 non-null    object 
 9    Inst Level.1         264 non-null    object 
 10   Grade Level Equiv.3  254 non-null    object 
 11   Accuracy (%)         218 non-null    float64
 12   Fluency (0-3)        232 non-null    float64
 13   Comp Within (0-3)    251 non-null    float64
 14   Comp Beyond (0-3)    249 non-null    float64
 15   Comp About           1

A couple of noticings:

* The column names are inconsistently formatted and need to be cleaned.
* Some numerical columns need to be casted as such.
* Since the Non-null counts are inconsistent, further investigation is needed.

In [5]:
# Removing spaces and special characters from column names 
df.columns = df.columns.str.strip()
df.columns = df.columns.str.replace('_',' ')

# Renaming Student Number & Homeroom
df.insert(1,'OSIS', df['Student  Number'])
df.insert(2,'Homeroom', df['Home room'])

# Creating list of columns to drop
drop_cols = ['Home room','Student  Number']

# Dropping columns
df.drop(drop_cols, axis=1, inplace=True)

In [6]:
# Renaming columns for easier readibility
df.insert(3,'BOY IP Level', df['BOY Indep Level'])
df.insert(4,'BOY IP Grade Equiv', df['Grade Level Equiv'])
df.insert(5,'BOY Instr Level', df['Inst Level'])
df.insert(6,'BOY Instr Grade Equiv', df['Grade Level Equiv.1'])
df.insert(7,'MOY IP Level', df['MOY Indep Level'])
df.insert(8,'MOY IP Grade Equiv', df['Grade Level Equiv.2'])
df.insert(9,'MOY Instr Level', df['Inst Level.1'])
df.insert(10,'MOY Instr Grade Equiv', df['Grade Level Equiv.3'])

In [7]:
# Dropping columns
drop_cols1 = ['BOY Indep Level','Grade Level Equiv','Inst Level','Grade Level Equiv.1','MOY Indep Level',
            'Grade Level Equiv.2','Inst Level.1','Grade Level Equiv.3']

df.drop(drop_cols1, axis=1, inplace=True)

In [8]:
# Checking for completion 
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 274 entries, 0 to 273
Data columns (total 17 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   OSIS                   274 non-null    float64
 1   Homeroom               274 non-null    object 
 2   Grade Level            274 non-null    float64
 3   BOY IP Level           273 non-null    object 
 4   BOY IP Grade Equiv     259 non-null    float64
 5   BOY Instr Level        187 non-null    object 
 6   BOY Instr Grade Equiv  274 non-null    object 
 7   MOY IP Level           271 non-null    object 
 8   MOY IP Grade Equiv     274 non-null    object 
 9   MOY Instr Level        264 non-null    object 
 10  MOY Instr Grade Equiv  254 non-null    object 
 11  Accuracy (%)           218 non-null    float64
 12  Fluency (0-3)          232 non-null    float64
 13  Comp Within (0-3)      251 non-null    float64
 14  Comp Beyond (0-3)      249 non-null    float64
 15  Comp A

The reading levels are documented in capital letters. As a precaution, I will make all independent and instructional levels capital.

In [9]:
# Capitalizing all reading levels

reading_level_cols =  ['BOY IP Level','BOY Instr Level', 'MOY IP Level', 'MOY Instr Level']

for col in reading_level_cols:
    df[col]=df[col].str.upper()

for i, row in df.iterrows():
    print(f"Index: {i}, {row['BOY IP Level']}, {row['MOY IP Level']}")

Index: 0, AA, AA
Index: 1, AA, AA
Index: 2, AA, AA
Index: 3, AA, AA
Index: 4, AA, AA
Index: 5, AA, A
Index: 6, AA, A
Index: 7, AA, A
Index: 8, AA, A
Index: 9, AA, A
Index: 10, AA, A
Index: 11, AA, B
Index: 12, AA, B
Index: 13, AA, B
Index: 14, AA, B
Index: 15, AA, D
Index: 16, AA, B
Index: 17, AA, C
Index: 18, AA, C
Index: 19, AA, C
Index: 20, AA, D
Index: 21, B, D
Index: 22, AA, D
Index: 23, AA, F
Index: 24,  , AA
Index: 25, AA, B
Index: 26,  , AA
Index: 27, AA, AA
Index: 28, AA, AA
Index: 29, AA,  
Index: 30, AA, AA
Index: 31, AA,  
Index: 32, AA, A
Index: 33, AA, A
Index: 34, AA, A
Index: 35, AA, A
Index: 36, AA,  
Index: 37, AA,  
Index: 38, AA,  
Index: 39, AA, A
Index: 40, AA, B
Index: 41, AA, B
Index: 42, AA, B
Index: 43, AA, C
Index: 44, AA, C
Index: 45, AA, C
Index: 46, AA, C
Index: 47, AA, C
Index: 48, AA, D
Index: 49, AA, D
Index: 50, AA, H
Index: 51, AA, H
Index: 52, AA, nan
Index: 53,  ,  
Index: 54, A, AA
Index: 55, AA, A
Index: 56, B, B
Index: 57, B, B
Index: 58, AA, B
I

In [10]:
# Getting summary of missing values
df.isna().sum()

OSIS                       0
Homeroom                   0
Grade Level                0
BOY IP Level               1
BOY IP Grade Equiv        15
BOY Instr Level           87
BOY Instr Grade Equiv      0
MOY IP Level               3
MOY IP Grade Equiv         0
MOY Instr Level           10
MOY Instr Grade Equiv     20
Accuracy (%)              56
Fluency (0-3)             42
Comp Within (0-3)         23
Comp Beyond (0-3)         25
Comp About               142
Total Comp                 0
dtype: int64

For BOY and MOY missing data, I will use the key provided with this analysis.

In [11]:
# importing BOY info
sheet_url = 'https://docs.google.com/spreadsheets/d/1ctQdLgknVDQ5BPCS-O6UihzK5kPbbVwI-svT3cWKWfw/edit#gid=1026408630'
url1 = sheet_url.replace('/edit#gid=', '/export?format=csv&gid=')        
df_BOY_levels = pd.read_csv(url1, header=1)
df_BOY_levels.head(27)

Independent Reading Level  Grade Equivalent                  K  \
0                         AA             -0.25     On Grade Level   
1                          A              0.00  Above Grade Level   
2                          B              0.33  Above Grade Level   
3                          C              0.67  Above Grade Level   
4                          D              1.00  Above Grade Level   
5                          E              1.16  Above Grade Level   
6                          F              1.33  Above Grade Level   
7                          G              1.50  Above Grade Level   
8                          H              1.67  Above Grade Level   
9                          I              1.83  Above Grade Level   
10                         J              2.00  Above Grade Level   
11                         K              2.33  Above Grade Level   
12                         L              2.67  Above Grade Level   
13                         M              3.00  Above Grade Level   
14                         N              3.33  Above Grade Level   
15                         O              3.67  Above Grade Level   
16                         P              4.00  Above Grade Level   
17                         Q              4.33  Above Grade Level   
18                         R              4.67  Above Grade Level   
19                         S              5.00  Above Grade Level   
20                         T              5.33  Above Grade Level   
21                         U              5.67  Above Grade Level   
22                         V              6.00  Above Grade Level   
23                         W              6.33  Above Grade Level   
24                         X              6.67  Above Grade Level   
25                         Y              7.00  Above Grade Level   
26                         Z              8.00  Above Grade Level   

                          1                        2                        3  \
0         Below Grade Level        Below Grade Level        Below Grade Level   
1         Below Grade Level        Below Grade Level        Below Grade Level   
2         Below Grade Level        Below Grade Level        Below Grade Level   
3   Approaching Grade Level        Below Grade Level        Below Grade Level   
4            On Grade Level        Below Grade Level        Below Grade Level   
5         Above Grade Level        Below Grade Level        Below Grade Level   
6         Above Grade Level        Below Grade Level        Below Grade Level   
7         Above Grade Level        Below Grade Level        Below Grade Level   
8         Above Grade Level        Below Grade Level        Below Grade Level   
9         Above Grade Level  Approaching Grade Level        Below Grade Level   
10        Above Grade Level           On Grade Level        Below Grade Level   
11        Above Grade Level        Above Grade Level        Below Grade Level   
12        Above Grade Level        Above Grade Level  Approaching Grade Level   
13        Above Grade Level        Above Grade Level           On Grade Level   
14        Above Grade Level        Above Grade Level        Above Grade Level   
15        Above Grade Level        Above Grade Level        Above Grade Level   
16        Above Grade Level        Above Grade Level        Above Grade Level   
17        Above Grade Level        Above Grade Level        Above Grade Level   
18        Above Grade Level        Above Grade Level        Above Grade Level   
19        Above Grade Level        Above Grade Level        Above Grade Level   
20        Above Grade Level        Above Grade Level        Above Grade Level   
21        Above Grade Level        Above Grade Level        Above Grade Level   
22        Above Grade Level        Above Grade Level        Above Grade Level   
23        Above Grade Level        Above Grade Level        Above Grade Level   
24        Above Grade Level        Above Grad

In [12]:
# Importing MOY info
sheet_url = 'https://docs.google.com/spreadsheets/d/1ctQdLgknVDQ5BPCS-O6UihzK5kPbbVwI-svT3cWKWfw/edit#gid=1314641425'
url2 = sheet_url.replace('/edit#gid=', '/export?format=csv&gid=')        
df_MOY_levels = pd.read_csv(url2, header=1)
df_MOY_levels.head(27)

Independent Reading Level  Grade Equivalent                        K  \
0                         AA             -0.25        Below Grade Level   
1                          A              0.00  Approaching Grade Level   
2                          B              0.33           On Grade Level   
3                          C              0.67        Above Grade Level   
4                          D              1.00        Above Grade Level   
5                          E              1.16        Above Grade Level   
6                          F              1.33        Above Grade Level   
7                          G              1.50        Above Grade Level   
8                          H              1.67        Above Grade Level   
9                          I              1.83        Above Grade Level   
10                         J              2.00        Above Grade Level   
11                         K              2.33        Above Grade Level   
12                         L              2.67        Above Grade Level   
13                         M              3.00        Above Grade Level   
14                         N              3.33        Above Grade Level   
15                         O              3.67        Above Grade Level   
16                         P              4.00        Above Grade Level   
17                         Q              4.33        Above Grade Level   
18                         R              4.67        Above Grade Level   
19                         S              5.00        Above Grade Level   
20                         T              5.33        Above Grade Level   
21                         U              5.67        Above Grade Level   
22                         V              6.00        Above Grade Level   
23                         W              6.33        Above Grade Level   
24                         X              6.67        Above Grade Level   
25                         Y              7.00        Above Grade Level   
26                         Z              8.00        Above Grade Level   

                          1                        2                        3  \
0         Below Grade Level        Below Grade Level        Below Grade Level   
1         Below Grade Level        Below Grade Level        Below Grade Level   
2         Below Grade Level        Below Grade Level        Below Grade Level   
3         Below Grade Level        Below Grade Level        Below Grade Level   
4         Below Grade Level        Below Grade Level        Below Grade Level   
5         Below Grade Level        Below Grade Level        Below Grade Level   
6   Approaching Grade Level        Below Grade Level        Below Grade Level   
7            On Grade Level        Below Grade Level        Below Grade Level   
8         Above Grade Level        Below Grade Level        Below Grade Level   
9         Above Grade Level        Below Grade Level        Below Grade Level   
10        Above Grade Level  Approaching Grade Level        Below Grade Level   
11        Above Grade Level           On Grade Level        Below Grade Level   
12        Above Grade Level        Above Grade Level        Below Grade Level   
13        Above Grade Level        Above Grade Level  Approaching Grade Level   
14        Above Grade Level        Above Grade Level           On Grade Level   
15        Above Grade Level        Above Grade Level        Above Grade Level   
16        Above Grade Level        Above Grade Level        Above Grade Level   
17        Above Grade Level        Above Grade Level        Above Grade Level   
18        Above Grade Level        Above Grade Level        Above Grade Level   
19        Above Grade Level        Above Grade Level        Above Grade Level   
20        Above Grade Level        Above Grade Level        Above Grade Level   
21        Above Grade Level        Above Grade Level        Above Grade Level   
22        Above Grade Level        Abov

BOY and MOY Independent Reading Levels & Grade Equivalents are identical.  I will create a dictionary with these level to assist with missing values.

In [13]:
# Dictionary 
alpha_keys = list(df_BOY_levels['Independent Reading Level'].unique())
alpha_values = list(df_BOY_levels['Grade Equivalent'].unique())

levels = dict(zip(alpha_keys, alpha_values))
levels

{'AA': -0.25,
 'A': 0.0,
 'B': 0.33,
 'C': 0.67,
 'D': 1.0,
 'E': 1.16,
 'F': 1.33,
 'G': 1.5,
 'H': 1.67,
 'I': 1.83,
 'J': 2.0,
 'K': 2.33,
 'L': 2.67,
 'M': 3.0,
 'N': 3.33,
 'O': 3.67,
 'P': 4.0,
 'Q': 4.33,
 'R': 4.67,
 'S': 5.0,
 'T': 5.33,
 'U': 5.67,
 'V': 6.0,
 'W': 6.33,
 'X': 6.67,
 'Y': 7.0,
 'Z': 8.0}

In [14]:
# Locating records with missing BOY IP level data
df.loc[(df['BOY IP Level'].isna())]

OSIS    Homeroom  Grade Level BOY IP Level  BOY IP Grade Equiv  \
67  238231914.0  Swarthmore          1.0          NaN                 NaN   

   BOY Instr Level BOY Instr Grade Equiv MOY IP Level MOY IP Grade Equiv  \
67             NaN                                  D                1.0   

   MOY Instr Level MOY Instr Grade Equiv  Accuracy (%)  Fluency (0-3)  \
67               E                  1.16           NaN            1.0   

    Comp Within (0-3)  Comp Beyond (0-3) Comp About  Total Comp  
67                1.0                2.0        NaN         3.0

This scholar is Swarthmore has data for the MOY assesment, but not the BOY assessment. I will assume this means the student was not assessed at the time. I will replace the BOY records with NE.

In [15]:
# Addressing missing data for student missing  
df['BOY IP Level'].fillna('NE', inplace=True)

df.loc[(df['BOY IP Level']=='NE'), 'BOY IP Grade Equiv']='NE'
df.loc[(df['BOY IP Level']=='NE'), 'BOY Instr Level']='NE'
df.loc[(df['BOY IP Level']=='NE'), 'BOY Instr Grade Equiv']='NE'

In [16]:
# Verifying accuracy
df.loc[(df['BOY IP Level'].isna())]

Empty DataFrame
Columns: [OSIS, Homeroom, Grade Level, BOY IP Level, BOY IP Grade Equiv, BOY Instr Level, BOY Instr Grade Equiv, MOY IP Level, MOY IP Grade Equiv, MOY Instr Level, MOY Instr Grade Equiv, Accuracy (%), Fluency (0-3), Comp Within (0-3), Comp Beyond (0-3), Comp About, Total Comp]
Index: []

In [17]:
# Locating records with missing BOY Grade Equivalent data
print(df.loc[df['BOY IP Grade Equiv'].isna()].shape)
df.loc[df['BOY IP Grade Equiv'].isna()]

(14, 17)


OSIS      Homeroom  Grade Level BOY IP Level BOY IP Grade Equiv  \
24   240415372.0      Syracuse          0.0                             NaN   
26   239668965.0      Syracuse          0.0                             NaN   
53   240905430.0        Howard          0.0                             NaN   
60   205888340.0    Swarthmore          1.0                             NaN   
73   237499462.0    Swarthmore          1.0                             NaN   
82   234292928.0    Swarthmore          1.0                             NaN   
83   218702165.0    Swarthmore          1.0                             NaN   
138  234743243.0      Williams          2.0                             NaN   
139  234411106.0      Williams          2.0                             NaN   
175  226787604.0      Tuskegee          3.0                             NaN   
189  233199898.0      Tuskegee          3.0                             NaN   
190  231197146.0      Tuskegee          3.0                             NaN   
195  203306691.0  Northwestern          3.0            H                NaN   
230  232795195.0    Georgetown          4.0                             NaN   

    BOY Instr Level BOY Instr Grade Equiv MOY IP Level MOY IP Grade Equiv  \
24                                                  AA                      
26                                                  AA                      
53                                                                          
60                                                   C               0.67   
73                                                   F               1.33   
82                                                                          
83                                                                          
138                                                                         
139                                                                         
175                                                  M                3.0   
189                                                                         
190                                                                         
195                                                  I                      
230                                                  Q               4.33   

    MOY Instr Level MOY Instr Grade Equiv  Accuracy (%)  Fluency (0-3)  \
24                A                   NaN           NaN            NaN   
26                A                   NaN           NaN            NaN   
53                                    NaN           NaN            NaN   
60                D                   1.0          99.0            2.0   
73                G                   1.5          94.0            1.0   
82                                    NaN           NaN            NaN   
83                                    NaN           NaN            NaN   
138                                   NaN           NaN            NaN   
139                                   NaN           NaN            NaN   
175               N                   NaN          99.0            2.0   
189                                   NaN           NaN            NaN   
190                                   NaN           NaN            NaN   
195               J                   NaN          98.0            2.0   
230               R                  4.67          99.0            3.0   

     Comp Within (0-3)  Comp Beyond (0-3) Comp About  Total Comp  
24                 NaN                NaN                    0.0  
26                 NaN                NaN                    0.0  
53                 NaN                NaN                    0.0  
60                 2.0                2.0        NaN         4.0  
73                 NaN                1.0        NaN         1.0  
82                 NaN                NaN                    0.0  
83                 NaN                NaN                    0.0  
138                NaN 

Upon further investigation, only one BOY IP Level was included (student with OSIS # 203306691). My assumption is that the other students do not have beginning year data because they were enrolled in school after school had started. I will first address the missing equivalency for the one student with data. Afterwards, I will fill those values with NE for "Not Enrolled." 

In [18]:
# Replacing grade equivalence for level H
df.loc[(df['OSIS']==203306691),'BOY IP Grade Equiv']=1.67

# Checking to see if change was made.
df.loc[df['OSIS']==203306691]

OSIS      Homeroom  Grade Level BOY IP Level BOY IP Grade Equiv  \
195  203306691.0  Northwestern          3.0            H               1.67   

    BOY Instr Level BOY Instr Grade Equiv MOY IP Level MOY IP Grade Equiv  \
195                                                  I                      

    MOY Instr Level MOY Instr Grade Equiv  Accuracy (%)  Fluency (0-3)  \
195               J                   NaN          98.0            2.0   

     Comp Within (0-3)  Comp Beyond (0-3) Comp About  Total Comp  
195                3.0                1.0        0.0         4.0

In [19]:
# Locating records with missing BOY Inst Level data
df.loc[df['BOY Instr Level'].isna()]

OSIS      Homeroom  Grade Level BOY IP Level BOY IP Grade Equiv  \
0    236563003.0      Syracuse          0.0           AA              -0.25   
1    238194344.0      Syracuse          0.0           AA              -0.25   
2    239980246.0      Syracuse          0.0           AA              -0.25   
3    240845834.0      Syracuse          0.0           AA              -0.25   
4    239980196.0      Syracuse          0.0           AA              -0.25   
..           ...           ...          ...          ...                ...   
213  237900477.0  Northwestern          3.0            P                4.0   
232  223455205.0    Georgetown          4.0            Q               4.33   
234  223474693.0    Georgetown          4.0            P                4.0   
239  205734916.0    Georgetown          4.0            R               4.67   
244  233255389.0    Georgetown          4.0            U               5.67   

    BOY Instr Level BOY Instr Grade Equiv MOY IP Level MOY IP Grade Equiv  \
0               NaN                                 AA              -0.25   
1               NaN                                 AA              -0.25   
2               NaN                                 AA              -0.25   
3               NaN                                 AA                      
4               NaN                                 AA                      
..              ...                   ...          ...                ...   
213             NaN                                  P                4.0   
232             NaN                                  Q               4.33   
234             NaN                                  Q               4.33   
239             NaN                                  R               4.67   
244             NaN                                  U               5.67   

    MOY Instr Level MOY Instr Grade Equiv  Accuracy (%)  Fluency (0-3)  \
0                 A                   0.0          88.0            1.0   
1                 A                   0.0          97.0            2.0   
2                 A                   0.0          33.0            2.0   
3                AA                 -0.25           NaN            3.0   
4                AA                 -0.25           NaN            2.0   
..              ...                   ...           ...            ...   
213               Q                  4.33          99.0            3.0   
232               S                   5.0          99.0            2.0   
234               R                  4.67          99.0            2.0   
239               S                   5.0          98.0            2.0   
244               V                   6.0          97.0            2.0   

     Comp Within (0-3)  Comp Beyond (0-3) Comp About  Total Comp  
0                  3.0                2.0        NaN         5.0  
1                  2.0                1.0        NaN         3.0  
2                  1.0                2.0        NaN         3.0  
3                  1.0                NaN        NaN         1.0  
4                  3.0                1.0        NaN         4.0  
..                 ...                ...        ...         ...  
213                2.0                2.0        1.0         5.0  
232                3.0                2.0        1.0         6.0  
234                3.0                2.0        1.0         6.0  
239                2.0                2.0        2.0         6.0  
244                1.0                1.0        0.0         2.0  

[86 rows x 17 columns]

<!-- mini-exploratory
The data shows that the instructional level is typically one level above the BOY Indep Level. 5.67 also appears to be highest level, which is equivalent to level U. -->

In [20]:
# Filling Instructional Level given BOY Independent Levels
import numpy as np

i = 0
for i, row in df.iterrows():
    for i in range(len(alpha_keys)-1):
        if pd.isnull(row['BOY Instr Level']):
            for level in alpha_keys:
                df.loc[(df['BOY IP Level']==alpha_keys[i]), 'BOY Instr Level']=alpha_keys[i+1]

In [21]:
# Checking for accuracy 
df.loc[df['BOY Instr Level'].isna()]

Empty DataFrame
Columns: [OSIS, Homeroom, Grade Level, BOY IP Level, BOY IP Grade Equiv, BOY Instr Level, BOY Instr Grade Equiv, MOY IP Level, MOY IP Grade Equiv, MOY Instr Level, MOY Instr Grade Equiv, Accuracy (%), Fluency (0-3), Comp Within (0-3), Comp Beyond (0-3), Comp About, Total Comp]
Index: []

In [22]:
# Spot-checking with previous student
df.loc[df['OSIS']==203306691]

OSIS      Homeroom  Grade Level BOY IP Level BOY IP Grade Equiv  \
195  203306691.0  Northwestern          3.0            H               1.67   

    BOY Instr Level BOY Instr Grade Equiv MOY IP Level MOY IP Grade Equiv  \
195               I                                  I                      

    MOY Instr Level MOY Instr Grade Equiv  Accuracy (%)  Fluency (0-3)  \
195               J                   NaN          98.0            2.0   

     Comp Within (0-3)  Comp Beyond (0-3) Comp About  Total Comp  
195                3.0                1.0        0.0         4.0

In [23]:
# Addressing BOY Instructional Level since all were missing, no conditionals were used.

i = 0
for level, equiv in levels.items():
    df.loc[(df['BOY Instr Level']==level), 'BOY Instr Grade Equiv'] = equiv
    
# Checking for accuracy    
df.tail(25)

OSIS   Homeroom  Grade Level BOY IP Level BOY IP Grade Equiv  \
249  229550850.0  Wisconsin          4.0            K               2.33   
250  227215274.0  Wisconsin          4.0            L               2.67   
251  209161520.0  Wisconsin          4.0            L               2.67   
252  225089911.0  Wisconsin          4.0            M                3.0   
253  230635724.0  Wisconsin          4.0            M                3.0   
254  230792921.0  Wisconsin          4.0            N               3.33   
255  227598208.0  Wisconsin          4.0            N               3.33   
256  232006445.0  Wisconsin          4.0            M                3.0   
257  219805173.0  Wisconsin          4.0            K               2.33   
258  208610493.0  Wisconsin          4.0            N               3.33   
259  231020033.0  Wisconsin          4.0            N               3.33   
260  231093345.0  Wisconsin          4.0            L               2.67   
261  213280321.0  Wisconsin          4.0            N               3.33   
262  231770355.0  Wisconsin          4.0            N               3.33   
263  227755410.0  Wisconsin          4.0            N               3.33   
264  232732271.0  Wisconsin          4.0            P                4.0   
265  232795203.0  Wisconsin          4.0            O               3.67   
266  230924680.0  Wisconsin          4.0            Q               4.33   
267  230560856.0  Wisconsin          4.0            R               4.67   
268  232298729.0  Wisconsin          4.0            P                4.0   
269  233376235.0  Wisconsin          4.0            Q               4.33   
270  232795112.0  Wisconsin          4.0            Q               4.33   
271  224448902.0  Wisconsin          4.0            R               4.67   
272  232836288.0  Wisconsin          4.0            R               4.67   
273  231019795.0  Wisconsin          4.0            R               4.67   

    BOY Instr Level BOY Instr Grade Equiv MOY IP Level MOY IP Grade Equiv  \
249               L                  2.67            K               2.33   
250               M                   3.0            L               2.67   
251               M                   3.0            L               2.67   
252               N                  3.33            N               3.33   
253               N                  3.33            N               3.33   
254               O                  3.67            N               3.33   
255               O                  3.67            N               3.33   
256               N                  3.33            N               3.33   
257               L                  2.67            N               3.33   
258               O                  3.67            N               3.33   
259               O                  3.67            N               3.33   
260               M                   3.0            O               3.67   
261               O                  3.67            O               3.67   
262               O                  3.67            O               3.67   
263               O                  3.67            P                4.0   
264               Q                  4.33            Q               4.33   
265               P                   4.0            Q               4.33   
266               R                  4.67            Q               4.33   
267               S                   5.0            R               4.67   
268               Q                  4.33            R               4.67   
269               R                  4.67            R               4.67   
270               R                  4.67            R               4.67   
271               S                   5.0            S                5.0   
272               S                   5.0            S                5.0   
273               S                   5.0            T               5.33   

    MOY Instr Level MOY Instr Gr

In [24]:
# Verifying there are no records with missing BOY Instr Grade Equiv
df.loc[df['BOY Instr Level'].isna()]

Empty DataFrame
Columns: [OSIS, Homeroom, Grade Level, BOY IP Level, BOY IP Grade Equiv, BOY Instr Level, BOY Instr Grade Equiv, MOY IP Level, MOY IP Grade Equiv, MOY Instr Level, MOY Instr Grade Equiv, Accuracy (%), Fluency (0-3), Comp Within (0-3), Comp Beyond (0-3), Comp About, Total Comp]
Index: []

In [25]:
# Completing cleaning process for MOY Data

In [26]:
########

In [27]:
# Locating records with missing MOY Indepent Level data
df.loc[df['MOY Indep Level'].isna()]

KeyError: 'MOY Indep Level'

In [ ]:
# Locating records with missing MOY data
print(df.loc[df['Grade Level Equiv.2'].isna()].shape)
df.loc[df['Grade Level Equiv.2'].isna()]

The Mid-Year IP Level is missing a few MOY Independent Levels. The MOY Independent Levels will be filled using these instructional levels.

In [ ]:
# Filling MOY Independent Levels given Instructional Level
i = 1
for i in range(len(alpha_keys)-1):
    if (df['MOY Indep Level'].isna().any()):
        for level in alpha_keys:
            df.loc[(df['Inst Level.1']==alpha_keys[i]), 'MOY Indep Level']=alpha_keys[i-1]
            i=+1

In [ ]:
# Checking for accuracy    
df.loc[df['MOY Indep Level'].isna()]

In [ ]:
df['MOY Indep Level'].shape

We now have 274 complete MOY Independent Levels. I will move on to investigate other columns.

In [ ]:
df.loc[df['Grade Level Equiv.2'].isna()]

In [ ]:
df.loc[df['Grade Level Equiv.2'].isna()].shape

In [ ]:
# Addressing Grade Level Equiv.2 for Instructional Level
for level, equiv in levels.items():
    df.loc[(df['MOY Indep Level']==level), 'Grade Level Equiv.2'] = equiv
    
# Checking for accuracy    
df.loc[df['Grade Level Equiv.2'].isna()].shape

In [ ]:
# Double checking NE students.

df.loc[df['BOY Indep Level']=="NE"]

In [ ]:
df.dropna(subset=['Grade Level Equiv.2'], inplace=True)

# Checking for accuracy
df.shape

In [ ]:
# Checking for accuracy    
df.loc[df['Grade Level Equiv.2'].isna()].shape

In [ ]:
df.shape

In [ ]:
# Next column: Instructional Level.
df.loc[df['Inst Level.1'].isna()].shape

In [ ]:
# Investigating 10 missing records.
df.loc[df['Inst Level.1'].isna()]

In [ ]:
# Filling MOY Instructional Level given MOY Independent Levels
i = 0
for i in range(len(alpha_keys)-1):
    for level in alpha_keys:
        df.loc[(df['MOY Indep Level']==alpha_keys[i]), 'Inst Level.1']=alpha_keys[i+1]
        i=+1

# Checking for completeness
df.loc[df['Inst Level.1'].isna()]

In [ ]:
# Filling Grade Level Equiv.3 given Instructional Levels
i = 0
for level, equiv in levels.items():
    df.loc[(df['Inst Level.1']==level), 'Grade Level Equiv.3'] = equiv
    
# Checking for completeness
df.loc[df['Grade Level Equiv.3'].isna()]

The data shows that students missing MOY data also were students who are also missing in the BOY. My assumption is that these students were registered, but didnot attend the school. I will drop these records from the data frame.

In [ ]:
df.dropna(subset=['Grade Level Equiv.3'], inplace=True)

In [ ]:
df['Grade Level Equiv.3'].isna().sum()

In [ ]:
df.info()

In [ ]:
# Removing students NE and blank MOY.


This first grader's BOY Independent Level was A and MOY Independent Level was Z. Such a high jump is highly unlikely. An investigation into these outliers are required.

In [ ]:
# Changing data to numerical type
df[['Grade Level Equiv','Grade Level Equiv.1','Grade Level Equiv.2','Grade Level Equiv.3']] = df[['Grade Level Equiv','Grade Level Equiv.1','Grade Level Equiv.2','Grade Level Equiv.3']].apply(pd.to_numeric, errors='coerce')

In [ ]:
# Investigating Outliers by creating a column showing the difference between BOY and MOY levels.
df.isna().sum()
df['BOY-MOY Growth'] = abs(df['Grade Level Equiv.2'] - df['Grade Level Equiv'])
df['BOY-MOY Growth']

Since this analysis compares student's reading levels from the beginning of the school to the middle of school year, I expect to see a few cases of regression, no growth, and atleast a half a year of growth in reading levels. According to F&P Instructional Level Expectations for Reading, students can grow 3-4 levels within a school year. For example, a first grader is expected to enter the grade reading on level E or above and be on level K or above by the end of the school year. For this part of the analysis, I will isolate students who grew more than two grade levels and investigate their growth.

In [28]:
for i, row in df.iterrows():
    print(f"Index: {i}, {row['BOY Indep Level']}")

KeyError: 'BOY Indep Level'

In [ ]:
for i, row in df.iterrows():
    print(f"Index: {i}, {row['BOY Indep Level']}, {row['MOY Indep Level']}")